### PR TITLE
Added performance benchmark for background platform channels

### PR DESCRIPTION
**Pull request by gaaclarke**
Benchmark for the work done to implement background platform channels in https://github.com/flutter/engine/pull/29147 (this PR can't land until that one does).

issue: https://github.com/flutter/flutter/issues/91635

Test exemption: This is a test.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Hixie** [comment]
@gaaclarke Since this is marked as a draft and hasn't been updated in a while, I'm going to close it for now just to get it off the review queue. Let me know if it should stay open. Thanks!

**Comment by gaaclarke** [comment]
Hey @Hixie, why are draft PR's being reviewed?  I thought by virtue of being a Draft, they aren't reviewed?  Otherwise, what's the point of the designation?

**Comment by Hixie** [comment]
Sorry I should have said "the open PR list" rather than "review queue". If you want are still using it and want it to remain open that's absolutely fine, we're just trying to tidy up PRs that are no longer on anyone's radar during the critical triage meeting (otherwise over the years we'll just have thousands of open abandoned draft PRs, heh).

**Comment by gaaclarke** [comment]
Okay, I would like to come back to this.  It's gotten deprioritized and I've started working on other stuff.  Having the draft PR helps me keep track of the work I've already done to implement this.

**Comment by skia-gold** [comment]
Gold has detected about 22 new digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/92130

**Comment by gaaclarke** [comment]
Here are the results when run locally:
```
I/flutter (27615): BasicMessageChannel/StandardMessageCodec/Flutter->Host/Small: 133.3 µs
I/flutter (27615): BasicMessageChannel/StandardMessageCodec/Flutter->Host/Large: 1629.0 µs
I/flutter (27615): BasicMessageChannel/BinaryCodec/Flutter->Host/Large: 123.0 µs
I/flutter (27615): BasicMessageChannel/BinaryCodec/Flutter->Host/1MB: 974.2 µs
I/flutter (27615): BasicMessageChannel/StandardMessageCodec/Flutter->Host/SmallParallel3: 72.7 µs
I/flutter (27615): BasicMessageChannel/StandardMessageCodec/Flutter->Host (background)/Small: 84.2 µs
I/flutter (27615): BasicMessageChannel/StandardMessageCodec/Flutter->Host (background)/SmallParallel3: 62.4 µs
```

That's a 45% reduction in roundtrip time when the messages are sent sequentially with a 4 byte payload, 15% reduction when the messages are parallelized to execute 3 at a time.

The hypothesis for the difference is that a bottleneck on the ui thread for handling messages equalizes the benefits of executing the handlers in parallel in background threads.

**Comment by gaaclarke** [comment]
friendly ping @dnfield

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 1]
```diff
@@ -59,8 +61,42 @@ Future<double> _runBasicStandardSmall(
...
 }
 
-Future<double> _runBasicStandardLarge(BasicMessageChannel<Object?> basicStandard,
-    List<Object?> largeBuffer, int count) async {
+class _Counter {
+  int count = 0;
+}
+
+void _runBasicStandardParallelRecurse(BasicMessageChannel<Object?> basicStandard,
+    _Counter counter, int count, Completer<int> completer, Object? payload) {
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:69]: nit: trailing comma and format

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 2]
```diff
@@ -59,8 +61,42 @@ Future<double> _runBasicStandardSmall(
...
+    completer.complete(counter.count);
+  } else if (counter.count < count) {
+    basicStandard.send(payload).then((Object? result){
+      _runBasicStandardParallelRecurse(basicStandard, counter, count, completer, payload);
+    });
+  }
+}
+
+Future<double> _runBasicStandardParallel(
+    BasicMessageChannel<Object?> basicStandard, int count, Object? payload, int parallel) async {
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:81]: ditto

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 3]
```diff
@@ -59,8 +61,42 @@ Future<double> _runBasicStandardSmall(
...
+  final Stopwatch watch = Stopwatch();
+  final Completer<int> completer = Completer<int>();
+  final _Counter counter = _Counter();
+  watch.start();
+  for( int i = 0; i < parallel; ++i) {
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:86]:
```suggestion
  for (int i = 0; i < parallel; ++i) {
```

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 4]
```diff
@@ -59,8 +61,42 @@ Future<double> _runBasicStandardSmall(
...
+  }
+  await completer.future;
+  watch.stop();
+  return watch.elapsedMicroseconds / count;
+}
+
+Future<double> _runBasicStandardLarge(
+    BasicMessageChannel<Object?> basicStandard,
+    List<Object?> largeBuffer,
+    int count) async {
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:99]: trailing comma

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 5]
```diff
@@ -101,18 +137,37 @@ Future<double> _runBasicBinary(BasicMessageChannel<ByteData> basicBinary,
   return watch.elapsedMicroseconds / count;
 }
 
+Future<void> _runTest({required Future<double> Function(int) test,
+    required BasicMessageChannel<Object?> resetChannel,
+    required BenchmarkResultPrinter printer,
+    required String description,
+    required String name,
+    required int numMessages}) async {
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:145]: trailing comma

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 6]
```diff
@@ -101,18 +137,37 @@ Future<double> _runBasicBinary(BasicMessageChannel<ByteData> basicBinary,
...
 Future<void> _runTests() async {
   if (kDebugMode) {
     throw Exception(
       "Must be run in profile mode! Use 'flutter run --profile'.",
     );
   }
 
-  const BasicMessageChannel<Object?> resetChannel = BasicMessageChannel<Object?>(
+  const BasicMessageChannel<Object?> resetChannel =
+      BasicMessageChannel<Object?>(
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:165]: Should not be on new line, same for L170

File: dev/benchmarks/platform_channels_benchmarks/lib/main.dart [hunk 7]
```diff
@@ -133,39 +188,70 @@ Future<void> _runTests() async {
...
+    resetChannel: resetChannel,
+    printer: printer,
+    description: 'BasicMessageChannel/StandardMessageCodec/Flutter->Host/SmallParallel3',
+    name: 'platform_channel_basic_standard_2host_small_parallel_3',
+    numMessages: numMessages,
   );
+  if (Platform.isAndroid) {
+    // Background platform channels aren't yet implemented for iOS.
+    const BasicMessageChannel<Object?> backgroundStandard =
+        BasicMessageChannel<Object?>(
```
Comment (dnfield) [inline dev/benchmarks/platform_channels_benchmarks/lib/main.dart:234]: ditto (no newline)

**Review by dnfield — APPROVED** [review]
LGTM with style nits

**Comment by gaaclarke** [comment]
Thanks Dan, I was having problems getting the formatter running.  I thought I fixed it before I pushed.

**Comment by gaaclarke** [comment]
Style updated.

**Pull request merged**